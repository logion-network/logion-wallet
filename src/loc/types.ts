import { LocRequestState } from "@logion/client";
import { UUID } from "@logion/node-api";
import { CallCallback } from "src/ClientExtrinsicSubmitter";

import { sha256HexFromString } from "../common/hash";

export type LocItemStatus = 'DRAFT' | 'PUBLISHED'

export type LocItemType = 'Data' | 'Document' | 'Linked LOC'

export interface LocItem {
    name: string,
    value: string,
    timestamp: string | null,
    type: LocItemType,
    submitter: string,
    status: LocItemStatus,
    nature?: string,
    target?: UUID,
    newItem: boolean,
    linkDetailsPath?: string,
}

export enum PublishStatus {
    NONE,
    START,
    PUBLISH_PENDING,
    PUBLISHING,
    PUBLISHED,
    ERROR
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem;
    itemType: 'Public Data' | 'Document' | 'Link';
    publishMutator: (current: LocRequestState, callback: CallCallback) => Promise<LocRequestState>;
}

export function toItemId(maybeHex: string): string | undefined {
    if(maybeHex.startsWith("0x") && maybeHex.length === 66) {
        return maybeHex;
    } else if(maybeHex.startsWith("0x") && maybeHex.length !== 66) {
        return undefined;
    } else {
        return `0x${sha256HexFromString(maybeHex)}`;
    }
}

export type ContributionMode = 'Requester' | 'VTP';
