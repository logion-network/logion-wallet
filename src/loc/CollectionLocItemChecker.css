.CollectionLocItemChecker {
    margin-top: 20px;
}

.CollectionLocItemChecker .item-id-input {
    display: inline-block;
    width: calc(100% - 32px);
}

.CollectionLocItemChecker .clear-button {
    cursor: pointer;
    margin-left: 7px;
}

.CollectionLocItemChecker .clear-button:hover {
    opacity: 0.5;
}

.CollectionLocItemChecker .Button {
    margin-left: 20px;
}

.CollectionLocItemChecker .IconTextRow .text {
    width: calc(100% - 80px);
}

.CollectionLocItemChecker .CheckResultFeedback {
    margin-top: 20px;
}

.CollectionLocItemChecker .CheckResultFeedback .label-positive {
    color: #37ad4b;
    font-weight: bold;
}

.CollectionLocItemChecker .CheckResultFeedback .label-negative {
    color: #ea1f46;
    font-weight: bold;
}

.CollectionLocItemChecker .CheckResultFeedback .Col {
    padding-right: 20px;
    min-width: 65px;
}

.CollectionLocItemChecker .CheckResultFeedback .url-header {
    font-weight: bold;
    margin-bottom: 0.5rem;
}

.CollectionLocItemChecker div.IconTextRow div.text div.FormGroup div div.Row div.Col {
    width: 50rem;
}

.CollectionLocItemChecker .url-copy-paste-container {
    display: flex;
}

.CollectionLocItemChecker .url-copy-paste-container .url-container {
    width: calc(100% - 20px);
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.CollectionLocItemChecker .url-copy-paste-container button {
    flex-shrink: 0;
}

.CollectionLocItemChecker .matched {
    border: 2px dashed #e6007a;
}

.CollectionLocItemChecker .buttons {
    display: flex;   
}

.CollectionLocItemChecker .ViewCertificateButton {
    padding-top: 0;
    padding-bottom: 3px;
    padding-left: 8px;
    padding-right: 4px;
}
