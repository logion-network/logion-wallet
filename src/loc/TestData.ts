import { LegalOfficerCase, UUID, Hash } from "@logion/node-api";
import { LocData, HashString } from "@logion/client";

export function buildLocRequest(locId: UUID, loc: LegalOfficerCase): LocData {
    return {
        ownerAddress: loc.owner,
        requesterAddress: loc.requesterAddress,
        requesterLocId: loc.requesterLocId,
        description: "Description",
        locType: loc.locType,
        createdOn: "",
        decisionOn: "",
        id: locId,
        status: loc.closed ? "CLOSED" : "OPEN",
        files: loc.files.map((locFile, index) => ({
            ...locFile,
            nature: HashString.fromValue(`File ${index}`),
            name: `File ${index}`,
            addedOn: "",
            published: false,
            restrictedDelivery: false,
            contentType: "text/plain",
            size: 42n,
            status: "DRAFT",
        })),
        metadata: loc.metadata.map((locFile, index) => ({
            ...locFile,
            name: HashString.fromValue(`Data ${index}`),
            value: HashString.fromValue(`Value ${index}`),
            addedOn: "",
            published: false,
            status: "DRAFT",
        })),
        links: loc.links.map((locFile, index) => ({
            ...locFile,
            name: `Link ${index}`,
            nature: HashString.fromValue(`Nature ${index}`),
            addedOn: "",
            target: locFile.id.toString(),
            published: false,
            status: "DRAFT",
        })),
        closed: false,
        verifiedIssuer: false,
        issuers: [],
    };
}
