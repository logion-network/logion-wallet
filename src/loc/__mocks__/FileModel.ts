export { getAllDeliveries, getAllCollectionDeliveries } from './FileModelMock';
