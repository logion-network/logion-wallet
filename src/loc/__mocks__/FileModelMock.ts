export let getAllDeliveries = jest.fn().mockResolvedValue({});
