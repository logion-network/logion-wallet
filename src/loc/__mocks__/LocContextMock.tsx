import { LocData, CollectionItem } from "@logion/client";
import { UUID, ValidAccountId } from "@logion/node-api";

import { LocItem } from "../LocItem";

let locState: any = {
    isLogionIdentity: () => false,
    isLogionData: () => true,
};

export function setLocState(value: any) {
    locState = value;
    if(value.data) {
        loc = {
            ...loc,
            ...locState.data(),
        }
    }
}

let locItems: LocItem[] = [];

export function setLocItems(items: LocItem[]) {
    locItems = items;
}

let locId = new UUID("aed4c6e4-979e-48ad-be6e-4bd39fb94762");

export function setLocId(id: UUID) {
    locId = id;
}

let loc: Partial<LocData> | null = {
    id: locId,
    locType: "Identity",
    requesterAccountId: ValidAccountId.polkadot("5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW"),
    userIdentity: {
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@logion.network",
        phoneNumber: "+1234",
    }
};

export function setLocRequest(request: Partial<LocData> | null) {
    loc = request;
}

export let refresh = jest.fn();

export let requestSof = jest.fn();

export function setRequestSof(fn: jest.Mock) {
    requestSof = fn;
}

export let deleteMetadata = jest.fn();

export let deleteLink = jest.fn();

let collectionItems: CollectionItem[] = [];

export function setCollectionItems(items: CollectionItem[]) {
    collectionItems = items;
}

async function mutateLocState(mutator: (state: any) => Promise<any>): Promise<void> {
    await mutator(locState);
}

export function useLocContext() {
    return {
        linkLoc: {

        },
        locId,
        loc,
        locItems,
        refresh,
        requestSof,
        deleteMetadata,
        deleteLink,
        checkResult: { result: "NONE" },
        locState,
        collectionItems,
        mutateLocState,
    };
}
