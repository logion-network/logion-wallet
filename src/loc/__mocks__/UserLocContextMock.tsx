import { UUID } from "@logion/node-api/dist/UUID";
import { LocData } from "@logion/client/dist/Loc";

import { LocRequest } from "../../common/types/ModelTypes";
import { LocItem } from "../types";

let locState: any;

export function setLocState(value: any) {
    locState = value;
}

export let refresh = jest.fn();

let requestSof = jest.fn();

export function setRequestSof(fn: jest.Mock) {
    requestSof = fn;
}

let requestSofOnCollection = jest.fn();

export function setRequestSofOnCollection(fn: jest.Mock) {
    requestSofOnCollection = fn;
}

let locId: UUID;

export function setLocId(id: UUID) {
    locId = id;
}

let loc: LocData | null = null;

export function setLoc(request: LocData | null) {
    loc = request;
}

let locItems: LocItem[] = [];

export function setLocItems(items: LocItem[]) {
    locItems = items;
}

export let deleteMetadata = jest.fn();

export let deleteFile = jest.fn();

export let addFile = jest.fn().mockResolvedValue(undefined);

export let addMetadata = jest.fn().mockResolvedValue(undefined);

export function useUserLocContext() {

    return {
        refresh,
        locState,
        requestSof,
        requestSofOnCollection,
        locId,
        loc,
        deleteMetadata,
        deleteFile,
        locItems,
        addFile,
        addMetadata,
    }
}
