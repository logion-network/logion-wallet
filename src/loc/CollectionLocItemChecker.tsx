import { AxiosInstance } from "axios";
import { useState, useCallback, useEffect } from "react";
import { Form } from "react-bootstrap";
import { ClosedCollectionLoc, CollectionItem } from "@logion/client";
import { UUID } from "@logion/node-api/dist/UUID";
import { getCollectionSize } from "@logion/node-api/dist/LogionLoc";

import { useLogionChain } from "../logion-chain";
import PolkadotFrame from "../common/PolkadotFrame";
import { useCommonContext } from "../common/CommonContext";
import Icon from "../common/Icon";
import IconTextRow from "../common/IconTextRow";
import FormGroup from "../common/FormGroup";
import Button from "../common/Button";
import { Row, Col } from "../common/Grid";
import CopyPasteButton from "../common/CopyPasteButton";
import { fullCollectionItemCertificate } from "../PublicPaths";

import "./CollectionLocItemChecker.css"
import StatementOfFactsButton from "./statement/StatementOfFactsButton";
import { toItemId, Viewer } from './types';
import StatementOfFactsRequestButton from "./statement/StatementOfFactsRequestButton";
import { useUserLocContext } from "./UserLocContext";
import { useLocContext } from "./LocContext";
import ItemFiles from "src/components/itemfiles/ItemFiles";
import { getAllDeliveries, ItemDeliveriesResponse } from "./FileModel";
import { useLegalOfficerContext } from "src/legal-officer/LegalOfficerContext";
import PagedTable, { getPage } from "src/components/pagedtable/PagedTable";
import { ActionCell, DateTimeCell } from "src/common/Table";
import ButtonGroup from "src/common/ButtonGroup";
import { useNavigate } from "react-router-dom";
import { dashboardCertificateRelativePath } from "src/RootPaths";
import CellWithCopyPaste from "src/components/table/CellWithCopyPaste";

export interface Props {
    locId: UUID;
    locOwner: string;
    collectionItem?: CollectionItem;
}

export type CheckResult = 'NONE' | 'POSITIVE' | 'NEGATIVE';

export function UserCollectionLocItemChecker(props: Props) {

    const { locId, collectionItem, locOwner } = props;
    const { locState, collectionItems } = useUserLocContext();
    const { client } = useLogionChain();
    const collection: ClosedCollectionLoc = locState as ClosedCollectionLoc;

    async function collectionItemFunction(actualId: string): Promise<CollectionItem | undefined> {
        const result = await collection.checkHash(actualId)
        return result.collectionItem;
    }

    if(!client) {
        return null;
    }

    return (
        <CollectionLocItemChecker
            viewer="User"
            locId={ locId }
            locOwner={ locOwner }
            collectionItem={ collectionItem }
            collectionSizeFunction={ () => collection.size() }
            collectionItemFunction={ collectionItemFunction }
            axiosFactory={ () => client.buildAxios(client.legalOfficers.find(legalOfficer => legalOfficer.address === locOwner)!) }
            collectionItems={ collectionItems }
        />)
}

export function LOCollectionLocItemChecker(props: Props) {

    const { locId, collectionItem, locOwner } = props;
    const { api } = useLogionChain();
    const { locState, collectionItems } = useLocContext();
    const { axios } = useLegalOfficerContext();
    const collection: ClosedCollectionLoc = locState as ClosedCollectionLoc;

    if (!api || !axios) {
        return null;
    }

    return (
        <CollectionLocItemChecker
            viewer="LegalOfficer"
            locId={ locId }
            locOwner={ locOwner }
            collectionItem={ collectionItem }
            collectionSizeFunction={ () => getCollectionSize({ api, locId }) }
            collectionItemFunction={ (actualId: string) => collection.getCollectionItem({ itemId: actualId }) }
            axiosFactory={ () => axios }
            collectionItems={ collectionItems }
        />)
}

interface LocalProps extends Props {
    viewer: Viewer;
    collectionSizeFunction: () => Promise<number | undefined>
    collectionItemFunction: (actualId: string) => Promise<CollectionItem | undefined>
    axiosFactory: () => AxiosInstance
    collectionItems: CollectionItem[]
}

function CollectionLocItemChecker(props: LocalProps) {

    const { locId, collectionSizeFunction, collectionItemFunction, locOwner, collectionItems } = props;
    const navigate = useNavigate();
    const { colorTheme } = useCommonContext();
    const { accounts, axiosFactory } = useLogionChain();
    const [ state, setState ] = useState<CheckResult>('NONE');
    const [ collectionSize, setCollectionSize ] = useState<number | undefined | null>(null);
    const [ itemId, setItemId ] = useState<string>("");
    const [ item, setItem ] = useState<CollectionItem>();
    const [ deliveries, setDeliveries ] = useState<ItemDeliveriesResponse>();
    const [ managedCheck, setManagedCheck ] = useState<{ itemId: string, active: boolean }>();
    const [ currentPageNumber, setCurrentPageNumber ] = useState(0);

    useEffect(() => {
        if (collectionSize === null) {
            collectionSizeFunction()
                .then(setCollectionSize)
        }
    }, [ collectionSizeFunction, collectionSize ])

    const checkData = useCallback(async () => {
        if (itemId) {
            const actualId = toItemId(itemId);
            if (actualId === undefined) {
                setState('NEGATIVE');
            } else {
                try {
                    const collectionItem = await collectionItemFunction(actualId);
                    if (collectionItem) {
                        setItem(collectionItem);
                        const deliveries = await getAllDeliveries(axiosFactory!(locOwner), { locId: locId.toString(), collectionItemId: itemId });
                        setDeliveries(deliveries);
                        setState('POSITIVE');
                    } else {
                        setState('NEGATIVE');
                    }
                } catch (e) {
                    console.log(e)
                    setState('NEGATIVE');
                }
            }
        }
    }, [ itemId, collectionItemFunction, axiosFactory, locId, locOwner ]);

    useEffect(() => {
        if (props.collectionItem && (
            !managedCheck
            || managedCheck.itemId !== props.collectionItem.id)) {

            setManagedCheck({
                itemId: props.collectionItem.id,
                active: true
            });
            setItemId(props.collectionItem.id);
            setItem(props.collectionItem);
            setState('POSITIVE');
        } else if (!props.collectionItem && managedCheck) {
            setManagedCheck(undefined);
            setItemId("");
            setState('NONE');
        }
    }, [ managedCheck, setManagedCheck, props.collectionItem ]);

    if (accounts?.current?.address === undefined) {
        return null;
    }

    return (
        <PolkadotFrame className="CollectionLocItemChecker" colorTheme={ colorTheme }>
            <IconTextRow
                icon={ <Icon icon={ { id: "polkadot_collection" } } width="45px" /> }
                text={ <>
                    <p className="text-title">Collection Item(s) recorded on logion blockchain</p>
                    <p>The Collection LOC content listed above shall cover all related Collection Items created using the logion import
                        tool and/or through the logion API by an external application approved between the Legal Officer and its client
                        under a process validated by the Legal Officer of the present Collection LOC.</p>
                    <p>To check if a Collection Item is covered by this Collection LOC and get its online public
                        certificate, just submit the related Collection Item ID in the input field below:</p>
                    <FormGroup
                        id="itemId"
                        noFeedback={ true }
                        control={
                            <Row>
                                <Col className={ managedCheck?.active ? "matched" : undefined }>
                                    <Form.Control
                                        type="text"
                                        value={ itemId }
                                        onChange={ value => {
                                            setState('NONE');
                                            if (managedCheck) {
                                                setManagedCheck({
                                                    itemId: managedCheck.itemId,
                                                    active: false
                                                });
                                            }
                                            setItemId(value.target.value);
                                        } }
                                        data-testid="item-id"
                                    />
                                </Col>
                                <Col className="buttons">
                                    <Button onClick={ checkData } disabled={ !itemId }><Icon
                                        icon={ { id: "search" } } /> Check Item ID</Button>
                                    {
                                        state === "POSITIVE" && item !== undefined &&
                                        (
                                            (props.viewer === 'LegalOfficer' &&
                                                <StatementOfFactsButton
                                                    item={ item }
                                                />
                                            ) ||
                                            (props.viewer === 'User' &&
                                                <StatementOfFactsRequestButton
                                                    itemId={ toItemId(itemId) }
                                                />
                                            )
                                        )
                                    }
                                </Col>
                            </Row>
                        }
                        colors={ colorTheme.frame }
                    />
                    <CheckResultFeedback
                        locId={ locId }
                        itemId={ toItemId(itemId) }
                        state={ state }
                        item={ item }
                        deliveries={ deliveries }
                        viewer={ props.viewer }
                        locOwner={ props.locOwner }
                        axiosFactory={ props.axiosFactory }
                    />
                </>
                }
            />
            <PagedTable
                fullSize={ collectionItems.length }
                currentPage={ getPage(collectionItems, currentPageNumber, 10) }
                columns={[
                    {
                        header: "Collection Item ID",
                        render: item => <CellWithCopyPaste content={ item.id } />,
                        align: "left",
                    },
                    {
                        header: "Timestamp",
                        render: item => <DateTimeCell dateTime={ item.addedOn } />,
                    },
                    {
                        header: "Certificate",
                        render: item => (
                            <ActionCell>
                                <ButtonGroup
                                    narrow={ true }
                                >
                                    <ViewCertificateButton
                                        url={ fullCollectionItemCertificate(locId, item.id) }
                                    />
                                    <CopyPasteButton
                                        className="medium"
                                        value={ fullCollectionItemCertificate(locId, item.id) }
                                    />
                                </ButtonGroup>
                            </ActionCell>
                        ),
                        width: "300px",
                    },
                    {
                        header: "Details",
                        render: item => (
                            <ActionCell>
                                <ButtonGroup>
                                    <Button
                                        onClick={ () => navigate(dashboardCertificateRelativePath("Collection", locId, item.id, props.viewer)) }
                                    >View</Button>
                                </ButtonGroup>
                            </ActionCell>
                        ),
                        width: "100px",
                    },
                ]}
                goToPage={ setCurrentPageNumber }
            />
        </PolkadotFrame>)
}

interface CheckResultProps {
    locId: UUID,
    itemId?: string,
    state: CheckResult,
    item?: CollectionItem,
    deliveries?: ItemDeliveriesResponse,
    viewer: Viewer,
    locOwner: string,
    axiosFactory: () => AxiosInstance,
}

function CheckResultFeedback(props: CheckResultProps) {
    const { locId, itemId, state, item, deliveries } = props;
    const { client } = useLogionChain();

    if(!client) {
        return null;
    }

    switch (state) {
        case "POSITIVE":
            const certificateUrl = fullCollectionItemCertificate(locId, itemId)
            return (
                <>
                    <Row className="CheckResultFeedback result-positive" id={ `feedback-${ state }` }>
                        <Col>
                            <p>
                                Check result: <span className="label-positive">positive</span><br />
                                The Collection Item - defined by the ID you submitted - is covered by the current Collection
                                LOC.
                            </p>
                        </Col>
                        <Col>
                            <Icon icon={ { id: "ok" } } height='45px' />
                        </Col>
                        <Col>
                            <div id="url-header" className="url-header">Certificate Public web address (URL) for the
                                data covered by this Collection LOC:
                            </div>
                            <div className="url-copy-paste-container">
                                <div className="url-container">
                                    <a href={ certificateUrl } target="_blank" rel="noreferrer">{ certificateUrl }</a>
                                </div>
                                <CopyPasteButton value={ certificateUrl } />
                            </div>
                        </Col>
                    </Row>
                    {
                        item !== undefined && item.files.length > 0 &&
                        <ItemFiles
                            item={ item }
                            deliveries={ deliveries }
                            withCheck={ props.viewer === "LegalOfficer" }
                        />
                    }
                </>
            )
        case "NEGATIVE":
            return (
                <Row className="CheckResultFeedback result-negative" id={ `feedback-${ state }` }>
                    <Col>
                        <p>
                            Check result: <span className="label-negative">negative</span><br />
                            The Collection Item - defined by the ID you submitted - has no match and is NOT covered by
                            the current<br />
                            Collection LOC. Please be careful and execute a deeper due diligence.
                        </p>
                    </Col>
                    <Col>
                        <Icon icon={ { id: "ko" } } height='45px' />
                    </Col>
                </Row>
            )
        case "NONE":
            return (
                <Row className="CheckResultFeedback result-none" children="" id={ `feedback-${ state }` } />
            )
    }
}

function ViewCertificateButton(props: { url: string }) {
    return (
        <Button className="ViewCertificateButton" onClick={ () => window.open(props.url, "_blank") }>
            <Icon icon={{ id:"view-certificate" }} width="16px" />
        </Button>
    );
}
