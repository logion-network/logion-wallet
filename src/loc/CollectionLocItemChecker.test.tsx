jest.mock("../common/CommonContext");
jest.mock("../logion-chain");
jest.mock("src/legal-officer/LegalOfficerContext");
jest.mock("./UserLocContext");
jest.mock("./FileModel");
jest.mock("./LocContext");
jest.mock("src/legal-officer/LegalOfficerContext");

import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { LocData, HashString } from "@logion/client";
import { UUID, Hash } from "@logion/node-api";

import { LOCollectionLocItemChecker, UserCollectionLocItemChecker, Props } from "./CollectionLocItemChecker";
import { DEFAULT_LEGAL_OFFICER } from "src/common/TestData";
import { clickByName } from "src/tests";
import { setLocState } from "./__mocks__/UserLocContextMock";
import { setLocState as setLoLocState, setLocRequest } from "./__mocks__/LocContextMock";
import { CollectionItem } from "@logion/client";
import { toItemId } from "./types";
import { It } from "moq.ts";
import { setupApiMock } from "src/__mocks__/LogionMock";

describe("CollectionLocItemChecker", () => {

    it("renders", () => {
        setupApiMock(api => {
            api.setup(instance => instance.queries.getCollectionSize(It.IsAny())).returnsAsync(1);
        });
        const result = render(<LOCollectionLocItemChecker collectionLoc={ locData } />);
        expect(result).toMatchSnapshot();
    })
})

describe("UserCollectionLocItemChecker", () => {

    it("shows positive match with item ID", async () => {
        setLocState(locState);
        await testWithItemId("test", UserCollectionLocItemChecker);
        await waitFor(() => expect(screen.getByText("positive")).toBeVisible());
    })

    it("shows negative match with wrong item ID", async () => {
        setLocState(locState);
        await testWithItemId("test2", UserCollectionLocItemChecker);
        await waitFor(() => expect(screen.getByText("negative")).toBeVisible());
    })
})

describe("LOCollectionLocItemChecker", () => {

    it("shows positive match with item ID", async () => {
        setLoLocState(locState);
        setLocRequest(locData);
        setupApiMock(api => {
            api.setup(instance => instance.queries.getCollectionSize(It.IsAny())).returnsAsync(1);
        });
        await testWithItemId("test", LOCollectionLocItemChecker);
        await waitFor(() => expect(screen.getByText("positive")).toBeVisible());
    })

    it("shows negative match with wrong item ID", async () => {
        setLoLocState(locState);
        setLocRequest(locData);
        await testWithItemId("test2", LOCollectionLocItemChecker);
        await waitFor(() => expect(screen.getByText("negative")).toBeVisible());
    })
})

const locId = new UUID("d97c99fd-9bcc-4f92-b9ea-b6be93abbbcd");
const itemId = toItemId("test")!;
const item = {
    id: itemId,
    description: HashString.fromValue("description"),
    files: [],
    termsAndConditions: [],
} as unknown as CollectionItem;
const locState = {
    size: () => Promise.resolve(1),
    checkHash: (checkHashItemId: Hash) => {
        if(checkHashItemId.equalTo(itemId)) {
            return Promise.resolve({ collectionItem: item });
        } else {
            return Promise.resolve({});
        }
    },
    getCollectionItem: (args: { itemId: Hash }) => {
        if(args.itemId.equalTo(itemId)) {
            return Promise.resolve(item);
        } else {
            return Promise.resolve();
        }
    }
};
const locData = {
    id: locId,
    ownerAddress: DEFAULT_LEGAL_OFFICER,
    metadata: [],
    files: [],
} as unknown as LocData;

async function testWithItemId(itemIdText: string, elementFunction: (props: Props) => JSX.Element | null) {
    render(elementFunction({ collectionLoc: locData }) || <div></div>);
    const itemIdInput = screen.getByTestId("item-id");
    await userEvent.type(itemIdInput, itemIdText);
    await clickByName(content => /Check Item ID/.test(content));
}
