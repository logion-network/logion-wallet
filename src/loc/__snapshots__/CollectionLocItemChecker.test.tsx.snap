// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionLocItemChecker renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Frame PolkadotFrame CollectionLocItemChecker"
        style="background-color: rgb(255, 255, 255); color: rgb(0, 0, 0); box-shadow: 0 0 25px #3b6cf41a; border: 2px solid #e6007a;"
      >
        <div
          class="IconTextRow"
        >
          <div
            class="icon"
          >
            <img
              alt="polkadot_collection"
              src="/assets/polkadot_collection.svg"
              width="45px"
            />
          </div>
          <div
            class="text"
          >
            <p
              class="text-title"
            >
              Collection Item(s) recorded within this LOC: 
              0
            </p>
            <p>
              All Collection Items created within this LOC - using the logion import tool and/or through the logion API by an external application - are governed by the public data and confidential documents recorded above.
            </p>
            <p>
              To check if a Collection Item is covered by this Collection LOC and get its online public certificate, just submit the related Collection Item ID in the input field below:
            </p>
            <div
              class="FormGroup"
              style="background-color: rgb(255, 255, 255);"
            >
              <style>
                
    .FormGroup .form-control,
    .FormGroup .form-control[readonly],
    .FormGroup .input-group .btn-primary {
        color: #000000;
    }
    
              </style>
              <div
                data-testid="itemIdGroup"
              >
                <div
                  class="Row"
                >
                  <div
                    class="Col"
                  >
                    <input
                      class="item-id-input form-control"
                      data-testid="item-id"
                      id="itemId"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="Col buttons"
                  >
                    <button
                      class="Button btn btn-primary"
                      disabled=""
                      type="button"
                    >
                      <img
                        alt="search"
                        src="/assets/search.svg"
                      />
                       Check Item ID
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Row CheckResultFeedback result-none"
              id="feedback-NONE"
            />
          </div>
        </div>
        <div
          class="PagedTable"
        >
          <div
            class="controls"
          >
            <span
              class="numbers"
            >
              1
              -
              0
               of 
              0
            </span>
            <span
              class="control left light disabled"
            >
              <img
                alt="left"
                height="10px"
                src="/assets/themes/light/left.svg"
              />
            </span>
            <span
              class="control right light disabled"
            >
              <img
                alt="right"
                height="10px"
                src="/assets/themes/light/right.svg"
              />
            </span>
          </div>
          <div
            class="Table"
          >
            <style>
              
            .Table .body .Row .Col.split-after::before {
                background-color: #152665;
            }
            
            </style>
            <div
              class="header"
              style="color: rgb(255, 255, 255); background-color: rgb(12, 22, 61);"
            >
              <div
                class="Row"
              >
                <div
                  class="Col"
                  style="text-align: left;"
                >
                  Collection Item ID
                </div>
                <div
                  class="Col"
                  style="text-align: left;"
                >
                  Description
                </div>
                <div
                  class="Col"
                  style="width: 200px; text-align: center;"
                >
                  Timestamp
                </div>
                <div
                  class="Col"
                  style="width: 160px; text-align: center;"
                >
                  Certificate
                </div>
                <div
                  class="Col"
                  style="width: 100px; text-align: center;"
                />
              </div>
            </div>
            <div
              class="body"
            >
              <div
                class="Row empty-data"
                style="color: rgb(255, 255, 255); background-color: rgb(21, 38, 101);"
              >
                <div
                  class="empty-message"
                >
                  No page to display
                </div>
              </div>
            </div>
          </div>
          <div
            class="controls"
          >
            <span
              class="numbers"
            >
              1
              -
              0
               of 
              0
            </span>
            <span
              class="control left light disabled"
            >
              <img
                alt="left"
                height="10px"
                src="/assets/themes/light/left.svg"
              />
            </span>
            <span
              class="control right light disabled"
            >
              <img
                alt="right"
                height="10px"
                src="/assets/themes/light/right.svg"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Frame PolkadotFrame CollectionLocItemChecker"
      style="background-color: rgb(255, 255, 255); color: rgb(0, 0, 0); box-shadow: 0 0 25px #3b6cf41a; border: 2px solid #e6007a;"
    >
      <div
        class="IconTextRow"
      >
        <div
          class="icon"
        >
          <img
            alt="polkadot_collection"
            src="/assets/polkadot_collection.svg"
            width="45px"
          />
        </div>
        <div
          class="text"
        >
          <p
            class="text-title"
          >
            Collection Item(s) recorded within this LOC: 
            0
          </p>
          <p>
            All Collection Items created within this LOC - using the logion import tool and/or through the logion API by an external application - are governed by the public data and confidential documents recorded above.
          </p>
          <p>
            To check if a Collection Item is covered by this Collection LOC and get its online public certificate, just submit the related Collection Item ID in the input field below:
          </p>
          <div
            class="FormGroup"
            style="background-color: rgb(255, 255, 255);"
          >
            <style>
              
    .FormGroup .form-control,
    .FormGroup .form-control[readonly],
    .FormGroup .input-group .btn-primary {
        color: #000000;
    }
    
            </style>
            <div
              data-testid="itemIdGroup"
            >
              <div
                class="Row"
              >
                <div
                  class="Col"
                >
                  <input
                    class="item-id-input form-control"
                    data-testid="item-id"
                    id="itemId"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="Col buttons"
                >
                  <button
                    class="Button btn btn-primary"
                    disabled=""
                    type="button"
                  >
                    <img
                      alt="search"
                      src="/assets/search.svg"
                    />
                     Check Item ID
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="Row CheckResultFeedback result-none"
            id="feedback-NONE"
          />
        </div>
      </div>
      <div
        class="PagedTable"
      >
        <div
          class="controls"
        >
          <span
            class="numbers"
          >
            1
            -
            0
             of 
            0
          </span>
          <span
            class="control left light disabled"
          >
            <img
              alt="left"
              height="10px"
              src="/assets/themes/light/left.svg"
            />
          </span>
          <span
            class="control right light disabled"
          >
            <img
              alt="right"
              height="10px"
              src="/assets/themes/light/right.svg"
            />
          </span>
        </div>
        <div
          class="Table"
        >
          <style>
            
            .Table .body .Row .Col.split-after::before {
                background-color: #152665;
            }
            
          </style>
          <div
            class="header"
            style="color: rgb(255, 255, 255); background-color: rgb(12, 22, 61);"
          >
            <div
              class="Row"
            >
              <div
                class="Col"
                style="text-align: left;"
              >
                Collection Item ID
              </div>
              <div
                class="Col"
                style="text-align: left;"
              >
                Description
              </div>
              <div
                class="Col"
                style="width: 200px; text-align: center;"
              >
                Timestamp
              </div>
              <div
                class="Col"
                style="width: 160px; text-align: center;"
              >
                Certificate
              </div>
              <div
                class="Col"
                style="width: 100px; text-align: center;"
              />
            </div>
          </div>
          <div
            class="body"
          >
            <div
              class="Row empty-data"
              style="color: rgb(255, 255, 255); background-color: rgb(21, 38, 101);"
            >
              <div
                class="empty-message"
              >
                No page to display
              </div>
            </div>
          </div>
        </div>
        <div
          class="controls"
        >
          <span
            class="numbers"
          >
            1
            -
            0
             of 
            0
          </span>
          <span
            class="control left light disabled"
          >
            <img
              alt="left"
              height="10px"
              src="/assets/themes/light/left.svg"
            />
          </span>
          <span
            class="control right light disabled"
          >
            <img
              alt="right"
              height="10px"
              src="/assets/themes/light/right.svg"
            />
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
